### PR TITLE
Convert http logs into structured ones

### LINKED ISSUE #27788: [FEATURE REQ] Use structured logging in core and http clients
We should review how we use logging in azure-core and azure-core-http-* and, where applicable, switch to structured logging

### PR DESCRIPTION
**Pull request by lmolkova**
Fixes #27788

### CODE REVIEW

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/DefaultRedirectStrategy.java [hunk 1]
```diff
@@ -94,8 +95,10 @@ && isValidRedirectCount(tryCount)
             && isAllowedRedirectMethod(httpResponse.getRequest().getHttpMethod())) {
             String redirectUrl = tryGetRedirectHeader(httpResponse.getHeaders(), getLocationHeader());
             if (redirectUrl != null && !alreadyAttemptedRedirectUrl(redirectUrl, attemptedRedirectUrls)) {
-                LOGGER.verbose("[Redirecting] Try count: {}, Attempted Redirect URLs: {}", tryCount,
-                    attemptedRedirectUrls.toString());
+                LOGGER.atVerbose()
+                    .addKeyValue(LoggingKeys.TRY_COUNT_KEY, tryCount)
+                    .addKeyValue(LoggingKeys.REDIRECT_URL_KEY, redirectUrl)
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/DefaultRedirectStrategy.java:100]: This no longer logs all the redirect URLs that were attempted, is this a logging behavior we want to change?

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/DefaultRedirectStrategy.java [hunk 2]
```diff
@@ -160,7 +166,11 @@ private boolean alreadyAttemptedRedirectUrl(String redirectUrl,
      */
     private boolean isValidRedirectCount(int tryCount) {
         if (tryCount >= getMaxAttempts()) {
-            LOGGER.error("Request has been redirected more than {} times.", getMaxAttempts());
+            LOGGER.atError()
+                .addKeyValue(LoggingKeys.TRY_COUNT_KEY, tryCount)
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/DefaultRedirectStrategy.java:170]: Does this need to include the try count as that is inferred by attempting more than the maximum limit

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/DefaultRedirectStrategy.java [hunk 3]
```diff
@@ -204,7 +217,10 @@ private boolean isValidRedirectStatusCode(int statusCode) {
     String tryGetRedirectHeader(HttpHeaders headers, String headerName) {
         String headerValue = headers.getValue(headerName);
         if (CoreUtils.isNullOrEmpty(headerValue)) {
-            LOGGER.error("Redirect url was null for header name: {}, request redirect was terminated.", headerName);
+            LOGGER.atError()
+                .addKeyValue("headerName", headerName)
+                .log("Redirect url was null, request redirect was terminated.");
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/DefaultRedirectStrategy.java:222]:
```suggestion
                .log("Redirect url header was null, request redirect was terminated.");
```

Thoughts?

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/logging/LoggingKeys.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.implementation.logging;
+
+public class LoggingKeys {
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/implementation/logging/LoggingKeys.java:6]: Mind adding Javadocs to these even though they're implementation

**Review by JonathanGiles — COMMENTED** [review]
Awesome work @lmolkova - thanks for doing this!

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/DefaultRedirectStrategy.java (hunk 1 above)
Comment (lmolkova) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/policy/DefaultRedirectStrategy.java:100]: good catch! I thought that Set.toString() would not work, but it does

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RetryPolicy.java [hunk 5]
```diff
@@ -142,18 +145,24 @@ private Mono<HttpResponse> attemptAsync(final HttpPipelineCallContext context, f
...
+                    LOGGER.atVerbose()
+                        .addKeyValue(LoggingKeys.TRY_COUNT_KEY, tryCount)
+                            .log("Error resume.", err);
                     return attemptAsync(context, next, originalHttpRequest, tryCount + 1)
                         .delaySubscription(retryStrategy.calculateRetryDelay(tryCount));
                 } else {
-                    LOGGER.info("Retry attempts have been exhausted after {} attempts.", tryCount, err);
+                    LOGGER.atVerbose()
+                        .addKeyValue(LoggingKeys.TRY_COUNT_KEY, tryCount)
+                        .log("Retry attempts have been exhausted.");
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RetryPolicy.java:165]: We should log this as info and also we should include the exception (`err`) in the log.
```suggestion
                    LOGGER.atInfo()
                        .addKeyValue(LoggingKeys.TRY_COUNT_KEY, tryCount)
                        .log("Retry attempts have been exhausted.", err);
```

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RetryPolicy.java [hunk 6]
```diff
@@ -142,18 +145,24 @@ private Mono<HttpResponse> attemptAsync(final HttpPipelineCallContext context, f
                     }
                 } else {
                     if (tryCount >= retryStrategy.getMaxRetries()) {
-                        LOGGER.info("Retry attempts have been exhausted after {} attempts.", tryCount);
+                        LOGGER.atVerbose()
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RetryPolicy.java:148]:
```suggestion
                        LOGGER.atInfo()
```

**Review by srnagar — APPROVED** [review]

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RetryPolicy.java (hunk 6 above)
Comment (lmolkova) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RetryPolicy.java:148]: great catch, thanks!

**Pull request merged**